cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,116 +1273,715 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+    ++current_iteration;
+    return bytes_copied;
+  }
 
-  // perform the copy.
-  copy_data(num_bufs, num_src_bufs, d_src_bufs, d_dst_bufs, d_dst_buf_info, stream);
+  bool has_more_copies() const { return current_iteration < num_iterations; }
 
-  // DtoH dst info (to retrieve null counts)
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    h_dst_buf_info, d_dst_buf_info, dst_buf_info_size, cudaMemcpyDefault, stream.value()));
+  rmm::device_uvector<dst_buf_info> d_batched_dst_buf_info;
+  rmm::device_uvector<offset_type> d_batch_offsets;
+  std::size_t total_size;
+  int num_iterations;
+  int current_iteration;
 
-  stream.synchronize();
+ private:
+  std::size_t starting_batch;
+  std::vector<std::size_t> h_num_buffs_per_iteration;
+  std::vector<std::size_t> h_size_of_buffs_per_iteration;
+};
+
+std::unique_ptr<chunk_iteration_state> make_chunk_iteration_state(
+  rmm::device_uvector<thrust::pair<std::size_t, std::size_t>> const& batches,
+  int num_bufs,
+  dst_buf_info* d_orig_dst_buf_info,
+  std::size_t const* const h_buf_sizes,
+  std::size_t num_partitions,
+  std::size_t user_buffer_size,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  rmm::device_uvector<offset_type> d_batch_offsets(num_bufs + 1, stream, mr);
+
+  auto buf_count_iter = cudf::detail::make_counting_transform_iterator(
+    0, [num_bufs, num_batches = num_batches_func{batches.begin()}] __device__(size_type i) {
+      return i == num_bufs ? 0 : num_batches(i);
+    });
 
-  // build the output.
-  std::vector<packed_table> result;
-  result.reserve(num_partitions);
-  std::vector<column_view> cols;
-  cols.reserve(num_root_columns);
-  auto cur_dst_buf_info = h_dst_buf_info;
-  for (std::size_t idx = 0; idx < num_partitions; idx++) {
-    // traverse the buffers and build the columns.
-    cur_dst_buf_info = build_output_columns(
-      input.begin(), input.end(), cur_dst_buf_info, std::back_inserter(cols), h_dst_bufs[idx]);
-
-    // pack the columns
-    cudf::table_view t{cols};
-    result.push_back(packed_table{
-      t,
-      packed_columns{
-        std::make_unique<std::vector<uint8_t>>(cudf::pack_metadata(
-          t, reinterpret_cast<uint8_t const*>(out_buffers[idx].data()), out_buffers[idx].size())),
-        std::make_unique<rmm::device_buffer>(std::move(out_buffers[idx]))}});
-
-    cols.clear();
+  thrust::exclusive_scan(rmm::exec_policy(stream, mr),
+                         buf_count_iter,
+                         buf_count_iter + num_bufs + 1,
+                         d_batch_offsets.begin(),
+                         0);
+
+  auto const num_batches_iter =
+    cudf::detail::make_counting_transform_iterator(0, num_batches_func{batches.begin()});
+  size_type const num_batches = thrust::reduce(
+    rmm::exec_policy(stream, mr), num_batches_iter, num_batches_iter + batches.size());
+
+  auto out_to_in_index = out_to_in_index_function{d_batch_offsets.begin(), num_bufs};
+
+  auto iter = thrust::make_counting_iterator(0);
```

Comment:
`const`

Reply:
updated this and other cases where the iterator is meant to not be mutable.